primitives/api/proc-macro/src/mock_impl_runtime_apis.rs

```diff
@@ -0,0 +1,364 @@
+				unimplemented!("Not required for testing!")
+			}
+		}
+	))
+}
+
+/// Auxialiry structure to fold a runtime api trait implementation into the expected format.
+///
+/// This renames the methods, changes the method parameters and extracts the error type.
+struct FoldRuntimeApiImpl<'a> {
+	/// The block type that is being used.
+	block_type: &'a TypePath,
+	/// The identifier of the trait being implemented.
+	impl_trait: &'a Ident,
+	/// Stores the error type that is being found in the trait implementation as associated type
+	/// with the name `Error`.
+	error_type: &'a mut Option<Type>,
+}
+
+impl<'a> Fold for FoldRuntimeApiImpl<'a> {
+	fn fold_impl_item_method(&mut self, mut input: syn::ImplItemMethod) -> syn::ImplItemMethod {
+		let block = {
+			let crate_ = generate_crate_access(HIDDEN_INCLUDES_ID);
+
+			let (param_names, param_types, error) = match extract_parameter_names_types_and_borrows(
+				&input.sig,
+				AllowSelfRefInParameters::YesButIgnore,
+			) {
+				Ok(res) => (
+					res.iter().map(|v| v.0.clone()).collect::<Vec<_>>(),
+					res.iter().map(|v| {
+						let ty = &v.1;
+						let borrow = &v.2;
+						quote!( #borrow #ty )
+					}).collect::<Vec<_>>(),
+					None
+				),
+				Err(e) => (Vec::new(), Vec::new(), Some(e.to_compile_error())),
+			};
+
+			let block_type = &self.block_type;
+
+			// Rewrite the input parameters.
+			input.sig.inputs = parse_quote! {
+				&self,
+				_: &#crate_::BlockId<#block_type>,
+				_: #crate_::ExecutionContext,
+				params: Option<( #( #param_types ),* )>,
+				_: Vec<u8>,
+			};
+
+			input.sig.ident = generate_method_runtime_api_impl_name(
+				&self.impl_trait,
+				&input.sig.ident,
+			);
+			let ret_type = return_type_extract_type(&input.sig.output);
+
+			// Generate the correct return type.
+			input.sig.output = parse_quote!(
+				-> std::result::Result<#crate_::NativeOrEncoded<#ret_type>, Self::Error>
+			);
+
+			let orig_block = input.block.clone();
+
+			// Generate the new method implementation that calls into the runtime.
+			parse_quote!(
+				{
+					// Get the error to the user (if we have one).
+					#error
+
+					let (#( #param_names ),*) = params
+						.expect("Mocked runtime apis don't support calling deprecated api versions");
+
+					let __fn_implementation__ = move || #orig_block;
+
+					Ok(#crate_::NativeOrEncoded::Native(__fn_implementation__()))
+				}
+			)
+		};
+
+		let mut input = fold::fold_impl_item_method(self, input);
+		// We need to set the block, after we modified the rest of the ast, otherwise we would
+		// modify our generated block as well.
+		input.block = block;
+		input
+	}
+
+	fn fold_impl_item(&mut self, input: ImplItem) -> ImplItem {
+		match input {
+			ImplItem::Type(ty) => {
+				if ty.ident == "Error" {
+					if let Some(error_type) = self.error_type {
+						if *error_type != ty.ty {
+							let error = Error::new(
+								ty.span(),
+								"Error type can not change between runtime apis",
+							);
```

Comment:
FIY, you can combine syn errors so that original `error_type` that differs from this `ty` is highlighted as well

something like this:
```rust
let error = Error::new(
	ty.span(),
	"Error type can not change between runtime apis",
 ).combine(Error::new(error_type.span(), "First Error type declared here"));
```